marda_extractors_api/__init__.py

```diff
@@ -0,0 +1,155 @@
+        subprocess.check_output(command)
+
+    def execute(self, file_type: str, file_path: Path, output_file: Path | None = None):
+        if file_type not in [d["id"] for d in self.entry["supported_filetypes"]]:
+            raise ValueError(
+                f"File type {file_type!r} not supported by {self.entry['id']!r}"
+            )
+
+        if self.entry["id"] == "yadg":
+            self.entry["usage"].append(
+                "python:yadg.extractors.extract({{ file_type }}, {{ file_path }})"
+            )
+
+        method, command = self.parse_usage(
+            self.entry["usage"], preferred_mode=self.preferred_mode
+        )
+
+        if output_file is None:
+            output_file = file_path.with_suffix(".json")
+
+        def apply_template_args(
+            command, file_type: str, file_path: Path, output_file: Path | None = None
+        ):
+            if method == SupportedExecutionMethod.CLI:
+                command = command.replace("{{ file_type }}", file_type)
+                command = command.replace("{{ file_path }}", str(file_path))
+                if output_file:
+                    command = command.replace("{{ output_file }}", str(output_file))
+            else:
+                command = command.replace("{{ file_type }}", f"{str(file_type)!r}")
+                command = command.replace("{{ file_path }}", f"{str(file_path)!r}")
+                if output_file:
+                    command = command.replace(
+                        "{{ output_file }}", f"{str(output_file)!r}"
+                    )
+
+            return command
+
+        command = apply_template_args(command, file_type, file_path, output_file)
+
+        if method == SupportedExecutionMethod.CLI:
+            output = self._execute_cli(command)
+            if not output_file.exists():
+                raise RuntimeError(f"Output file {output_file} does not exist")
+            print(f"Wrote output to {output_file}")
+
+        elif method == SupportedExecutionMethod.PYTHON:
+            output = self._execute_python(command)
+
+        return output
+
+    def _execute_cli(self, command):
+        return subprocess.check_output(command)
+
+    def _execute_python(self, command):
+        from importlib import import_module
+
+        module = ".".join(command.split("(")[0].split(".")[0:-1])
+
+        function = command.split("(")[0].split(".")[-1]
+        extractor_module = import_module(module)
+        # Treat contents of first brackets as arguments
+        args = [
+            d.strip().strip("'") for d in command.split("(")[1].split(")")[0].split(",")
+        ]
+        return getattr(extractor_module, function)(*args)
+
+    @staticmethod
+    def parse_usage(
+        usage: list[str],
+        preferred_mode: SupportedExecutionMethod = SupportedExecutionMethod.PYTHON,
+    ) -> tuple[SupportedExecutionMethod, str]:
+        """Parse e.g., 'cli: yadg extract {{ file_type }} {{ file_path }} {{ output_file }}'."""
```

Comment:
Perhaps we should keep our nomenclature consistent. `{{ file_type }}` is OK, but I'd prefer `{{ input_file }}` and `{{ output_file }}`.

Reply:
Yes, we should properly define these... perhaps in the meeting today? There was also a good comment from Logan (won't tag to avoid spamming him) in #4 about a context dictionary for arbitrary params.